pipeline_dp/dp_engine.py

```diff
@@ -288,6 +292,37 @@ def _add_empty_public_partitions(self, col, public_partitions,
             col, empty_accumulators,
             "Join public partitions with partitions from data")
 
+    def _bound_per_user_contributions(self, col, max_contributions: int,
+                                      aggregator_fn):
+        """Bounds the contribution by privacy_id in and cross partitions.
+
+        Args:
+          col: collection, with types of each element: (privacy_id,
+            partition_key, value).
+          max_contributions: maximum number of records that one privacy id can
+            contribute.
+          aggregator_fn: function that takes a list of values and returns an
+            aggregator object which handles all aggregation logic.
+
+        return: collection with elements ((privacy_id, partition_key),
+              accumulator).
+        """
+        # Max contributions bounding
+        col = self._backend.sample_fixed_per_key(col, max_contributions,
+                                                 "Sample per privacy_id")
+        self._add_report_stage(
+            f"max contribution bounding: randomly selected not "
+            f"more than {max_contributions} contributions")
+        col = self._backend.map_tuple(
```

Comment:
In order to be consistent with `_contribution_bounding` we shoud aggregate per (privacy_id, partition), nicely that we already grouped per `privacy_id` (and sumsampled), we can assumet that `max_contributions` is small enough to make everything in memory for one `(privacy_id, )`. 
Namely here we can use `map_values` with `fn` which takes a list ` [(partition_key, value)]` and this `fn` groups by `partition_key` (e.g. with dictionaries) and then applies `aggregator_fn` for values per each.

Does it make sense?

Reply:
I don't think I fully understand this. Possible to discuss this sometime?

Reply:
Thanks for the pointers. Have fixed the implementation and have also added test cases to verify this.